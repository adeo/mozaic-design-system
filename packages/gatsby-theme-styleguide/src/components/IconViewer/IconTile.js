import React, { useState } from 'react'
import Copy from '../Copy'
import * as styles from './iconviewer.module.css'

const buildIconData = (name, icons) => {
  let dataIcon = {
    name: { name },
    currentSize: '48px',
    sizes: {
      '16px': {
        url: null,
        component: null,
        fileName: null,
        fullName: null,
        iconName: null,
      },
      '24px': {
        url: null,
        component: null,
        fileName: null,
        fullName: null,
        iconName: null,
      },
      '32px': {
        url: null,
        component: null,
        fileName: null,
        fullName: null,
        iconName: null,
      },
      '48px': {
        url: null,
        component: null,
        fileName: null,
        fullName: null,
        iconName: null,
      },
      '64px': {
        url: null,
        component: null,
        fileName: null,
        fullName: null,
        iconName: null,
      },
    },
  }

  Object.keys(dataIcon.sizes).forEach((size) => {
    if (!icons[size]) {
      return false
    } else {
      dataIcon.sizes[size].component = icons[size].componentName || null
      dataIcon.sizes[size].fileName = icons[size].fileName
      dataIcon.sizes[size].iconName = icons[size].iconName
    }

    if (!icons[dataIcon.currentSize]) {
      dataIcon.currentSize = Object.keys(icons).find((size) => {
        return icons[size] != null
      })
    }

    return true
  })

  return dataIcon
}

const createWebComponentName = (str) =>
  str.replace(/_/g, '-').replace('.svg', '').toLowerCase()

const TechnicalName = ({ keyName, value }) => (
  <p className={styles.technicalName}>
    {keyName}
    <br />
    {value}
  </p>
)

const IconTile = (props) => {
  const dataIcon = buildIconData(props.name, props.icons)
  const [currentSize, setCurrentSize] = useState(dataIcon.currentSize)
  const [allIcon, setAllIcon] = useState(dataIcon.sizes)

  function changeIconSize(size) {
    setCurrentSize(size)
  }

  return (
    <li className={styles.detail}>
      <p className={styles.iconName}>{props.name}</p>
      <div className={styles.iconWrapper}>
        {props.type === 'monochrom' ? (
          <img
            src={`/icons/monochrom/${allIcon[currentSize].fileName}`}
            alt={allIcon[currentSize].fileName}
            height={currentSize}
            width={currentSize}
            loading="lazy"
          />
        ) : (
          <img
            src={`/icons/color/${allIcon[currentSize].fileName}`}
            alt={allIcon[currentSize].fileName}
            height={currentSize}
            width={currentSize}
            loading="lazy"
          />
        )}
      </div>
      <ul className={styles.sizeList}>
        {Object.keys(dataIcon.sizes).map((size) => (
          <li
            key={`${props.name}-${size}`}
            className={`
              ${styles.sizeListItem} 
              ${currentSize === size ? styles.isActive : ''}
              ${!allIcon[size].fileName ? styles.isDisabled : ''}
            `}
          >
            <button
              type="button"
              className={styles.sizeButton}
              onClick={(e) => changeIconSize(size, e)}
              disabled={allIcon[size].fileName ? false : true}
            >
              {size}
            </button>
          </li>
        ))}
      </ul>
      {allIcon[currentSize].iconName && (
        <Copy
          value={allIcon[currentSize].iconName}
          children={
            <TechnicalName
              keyName="Name: "
              value={allIcon[currentSize].iconName}
            />
          }
        />
      )}
      {allIcon[currentSize].component && (
        <Copy
          value={allIcon[currentSize].component}
          children={
            <TechnicalName
              keyName="React/Svelte/Vue:"
              value={allIcon[currentSize].component}
            />
          }
        />
      )}
      {allIcon[currentSize].fileName && (
        <Copy
          value={createWebComponentName(allIcon[currentSize].fileName)}
          children={
            <TechnicalName
              keyName="Web Component: "
              value={createWebComponentName(allIcon[currentSize].fileName)}
            />
          }
        />
      )}
      <Copy
        value={allIcon[currentSize].fileName}
        children={
          <TechnicalName
            keyName="Svg file : "
            value={allIcon[currentSize].fileName}
          />
        }
      />
    </li>
  )
}

export default IconTile
