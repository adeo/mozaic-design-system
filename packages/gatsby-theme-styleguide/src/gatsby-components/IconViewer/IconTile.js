import React, { Component } from 'react'
import styled from 'styled-components'
import Copy from '../Copy'
import * as icons from '@mozaic-ds/icons/react'

const IconDetail = styled.li`
  flex: 0 1 30.8%;
  list-style: none;
  border: 1px solid #eee;
  box-sizing: border-box;
  border-radius: 4px;
  margin: 0.5rem;
`

const IconWrapper = styled.div`
  text-align: center;
  height: 104px;
  line-height: 3.75;
  display: flex;

  img,
  svg {
    margin: auto;
  }
`

const SizesList = styled.ul`
  display: flex;
  margin: 0 !important;
  border-top: 1px solid #eee;
  border-bottom: 1px solid #eee;
`
const SizeItem = styled.li`
  flex: 0 1 20%;
  list-style: none;
  text-align: center;
  font-size: 0.65rem;
  line-height: 1.62rem;

  background: #f2f2f2;
  color: darkgrey;

  &:not(:first-child) {
    border-left: 1px solid #eee;
  }
`

const SizeItemAvail = styled(SizeItem)`
  background: ${({ isActive }) => (isActive ? `#78be20` : `none`)};
  color: ${({ isActive }) => (isActive ? `white !important` : `inherit`)};
`

export const SizeItemBtn = styled.button`
  box-shadow: none;
  margin: 0;
  padding: 0;
  border: none;
  font-size: 0.65rem;
  line-height: 1;
  background: none;
  cursor: pointer;
  color: inherit;
  padding: 0.5rem 0.25rem;
`

const IconName = styled.p`
  padding: 1rem;
  border-bottom: 1px solid #eee;
  font-size: 0.75rem;
  margin: 0;
  font-weight: 600;
  word-break: break-all;
  height: 40px;
`

const IconSizeName = styled.p`
  padding: 0 3px;
  margin: 3px;
  line-height: 1.25;
  display: block;
  word-break: break-all;
  position: relative;
`

export default class IconTile extends Component {
  constructor(props) {
    super(props)

    this.dataIcon = this.buildIconData(props.name, props.icons)

    this.state = {
      currentSize: this.dataIcon.currentSize,
      allIcon: this.dataIcon.sizes,
    }
  }

  changeIconSize(size) {
    this.setState({ currentSize: size })
  }

  buildIconData(name, icons) {
    let dataIcon = {
      name: { name },
      currentSize: '48px',
      sizes: {
        '16px': {
          url: null,
          component: null,
          fileName: null,
          fullName: null,
          iconName: null,
        },
        '24px': {
          url: null,
          component: null,
          fileName: null,
          fullName: null,
          iconName: null,
        },
        '32px': {
          url: null,
          component: null,
          fileName: null,
          fullName: null,
          iconName: null,
        },
        '48px': {
          url: null,
          component: null,
          fileName: null,
          fullName: null,
          iconName: null,
        },
        '64px': {
          url: null,
          component: null,
          fileName: null,
          fullName: null,
          iconName: null,
        },
      },
    }

    Object.keys(dataIcon.sizes).forEach(size => {
      if (!icons[size]) {
        return false
      } else {
        dataIcon.sizes[size].component = icons[size].componentName || null
        dataIcon.sizes[size].fileName = icons[size].fileName
        dataIcon.sizes[size].iconName = icons[size].iconName
      }

      if (!icons[dataIcon.currentSize]) {
        dataIcon.currentSize = Object.keys(icons).find(size => {
          return icons[size] != null
        })
      }

      return true
    })

    return dataIcon
  }

  render() {
    const Icon = icons[this.state.allIcon[this.state.currentSize].component]

    return (
      <IconDetail>
        <IconName>{this.props.name}</IconName>
        <IconWrapper>
          {this.props.type === 'monochrom' ? (
            <Icon fill="#454545" size={this.state.currentSize} />
          ) : (
            <img
              src={`/colorsvg/${
                this.state.allIcon[this.state.currentSize].fileName
              }`}
              width={this.state.currentSize}
              height={this.state.currentSize}
              alt={this.state.allIcon[this.state.currentSize].fileName}
            />
          )}
        </IconWrapper>
        <SizesList>
          {Object.keys(this.dataIcon.sizes).map(size =>
            this.state.allIcon[size].fileName ? (
              <SizeItemAvail
                key={`${this.props.name}-${size}`}
                isActive={this.state.currentSize === size}
              >
                <SizeItemBtn
                  type="button"
                  onClick={e => this.changeIconSize(size, e)}
                >
                  {size}
                </SizeItemBtn>
              </SizeItemAvail>
            ) : (
              <SizeItem key={`${this.props.name}-${size}`}>{size}</SizeItem>
            )
          )}
        </SizesList>
        {this.state.allIcon[this.state.currentSize].iconName && (
          <Copy
            value={this.state.allIcon[this.state.currentSize].iconName}
            children={
              <TechnicalName
                keyName="Name : "
                value={this.state.allIcon[this.state.currentSize].iconName}
              />
            }
          />
        )}
        {this.state.allIcon[this.state.currentSize].component && (
          <Copy
            value={this.state.allIcon[this.state.currentSize].component}
            children={
              <TechnicalName
                keyName="React/vue : "
                value={this.state.allIcon[this.state.currentSize].component}
              />
            }
          />
        )}
        <Copy
          value={this.state.allIcon[this.state.currentSize].fileName}
          children={
            <TechnicalName
              keyName="Svg file : "
              value={this.state.allIcon[this.state.currentSize].fileName}
            />
          }
        />
      </IconDetail>
    )
  }
}

const TechnicalName = ({ keyName, value }) => (
  <IconSizeName>
    {keyName}
    <br />
    {value}
  </IconSizeName>
)
