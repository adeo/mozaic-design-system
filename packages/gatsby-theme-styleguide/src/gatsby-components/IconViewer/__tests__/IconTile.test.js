import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import IconTile, { SizeItemBtn } from '../IconTile'

const ColorProps = {
  type: 'color',
  name: 'StoreLocation--Colors',
  icons: {
    '16px': null,
    '24px': null,
    '32px': null,
    '48px': {
      cat: 'Store',
      fileName: 'Store_StoreLocation--Colors_48px.svg',
      name: 'StoreLocation--Colors',
      size: '48px',
    },
    '64px': {
      cat: 'Store',
      fileName: 'Store_StoreLocation--Colors_64px.svg',
      name: 'StoreLocation--Colors',
      size: '64px',
    },
  },
}

const ColorPropsNo48 = {
  type: 'color',
  name: 'StoreLocation--Colors',
  icons: {
    '16px': null,
    '24px': null,
    '32px': {
      cat: 'Store',
      fileName: 'Store_StoreLocation--Colors_48px.svg',
      name: 'StoreLocation--Colors',
      size: '48px',
    },
    '48px': null,
    '64px': {
      cat: 'Store',
      fileName: 'Store_StoreLocation--Colors_64px.svg',
      name: 'StoreLocation--Colors',
      size: '64px',
    },
  },
}

const monochromProps = {
  type: 'monochrom',
  name: 'Mobile',
  icons: {
    '16px': null,
    '24px': null,
    '32px': null,
    '48px': {
      cat: 'Device',
      componentName: 'Mobile48',
      fileName: 'Device_Mobile_48px.svg',
      name: 'Mobile',
      size: '48px',
    },
    '64px': null,
  },
}

describe('<IconTile />', () => {
  it('render corectly icon colors', () => {
    const tree = renderer.create(<IconTile {...ColorProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render corectly monochrom icon', () => {
    const tree = renderer.create(<IconTile {...monochromProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('change the icon when you click on another size', () => {
    const Tile = mount(<IconTile {...ColorProps} />)
    expect(Tile.find(SizeItemBtn).length).toBe(2)
    Tile.find(SizeItemBtn)
      .at(1)
      .simulate('click')
    expect(Tile.state().currentSize).toBe('64px')
  })

  it('display another size when default size (48px) is not available', () => {
    const Tile = mount(<IconTile {...ColorPropsNo48} />)
    expect(Tile.state().currentSize).toBe('32px')
  })
})
