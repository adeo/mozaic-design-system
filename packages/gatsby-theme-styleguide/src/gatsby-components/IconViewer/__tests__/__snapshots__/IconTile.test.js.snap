// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconTile /> render corectly icon colors 1`] = `
.c8 {
  width: 100%;
  padding: 3px 5px;
  font-size: 10px;
  color: #999;
  line-height: 2em;
  cursor: pointer;
  position: relative;
  box-sizing: border-box;
}

.c8:not(:last-child) {
  border-bottom: solid 1px #f5f5f5;
}

.c0 {
  -webkit-flex: 0 1 30.8%;
  -ms-flex: 0 1 30.8%;
  flex: 0 1 30.8%;
  list-style: none;
  border: 1px solid #eee;
  box-sizing: border-box;
  border-radius: 4px;
  margin: 0.5rem;
}

.c2 {
  text-align: center;
  height: 104px;
  line-height: 3.75;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 img,
.c2 svg {
  margin: auto;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 !important;
  border-top: 1px solid #eee;
  border-bottom: 1px solid #eee;
}

.c4 {
  -webkit-flex: 0 1 20%;
  -ms-flex: 0 1 20%;
  flex: 0 1 20%;
  list-style: none;
  text-align: center;
  font-size: 0.65rem;
  line-height: 1.62rem;
  background: #f2f2f2;
  color: darkgrey;
}

.c4:not(:first-child) {
  border-left: 1px solid #eee;
}

.c5 {
  -webkit-flex: 0 1 20%;
  -ms-flex: 0 1 20%;
  flex: 0 1 20%;
  list-style: none;
  text-align: center;
  font-size: 0.65rem;
  line-height: 1.62rem;
  background: #f2f2f2;
  color: darkgrey;
  background: #78be20;
  color: white !important;
}

.c5:not(:first-child) {
  border-left: 1px solid #eee;
}

.c7 {
  -webkit-flex: 0 1 20%;
  -ms-flex: 0 1 20%;
  flex: 0 1 20%;
  list-style: none;
  text-align: center;
  font-size: 0.65rem;
  line-height: 1.62rem;
  background: #f2f2f2;
  color: darkgrey;
  background: none;
  color: inherit;
}

.c7:not(:first-child) {
  border-left: 1px solid #eee;
}

.c6 {
  box-shadow: none;
  margin: 0;
  padding: 0;
  border: none;
  font-size: 0.65rem;
  line-height: 1;
  background: none;
  cursor: pointer;
  color: inherit;
  padding: 0.5rem 0.25rem;
}

.c1 {
  padding: 1rem;
  border-bottom: 1px solid #eee;
  font-size: 0.75rem;
  margin: 0;
  font-weight: 600;
  word-break: break-all;
  height: 40px;
}

.c9 {
  padding: 0 3px;
  margin: 3px;
  line-height: 1.25;
  display: block;
  word-break: break-all;
  position: relative;
}

<li
  className="c0"
>
  <p
    className="c1"
  >
    StoreLocation--Colors
  </p>
  <div
    className="c2"
  >
    <img
      alt="Store_StoreLocation--Colors_48px.svg"
      height="48px"
      src="/colorsvg/Store_StoreLocation--Colors_48px.svg"
      width="48px"
    />
  </div>
  <ul
    className="c3"
  >
    <li
      className="c4"
    >
      16px
    </li>
    <li
      className="c4"
    >
      24px
    </li>
    <li
      className="c4"
    >
      32px
    </li>
    <li
      className="c5"
    >
      <button
        className="c6"
        onClick={[Function]}
        type="button"
      >
        48px
      </button>
    </li>
    <li
      className="c7"
    >
      <button
        className="c6"
        onClick={[Function]}
        type="button"
      >
        64px
      </button>
    </li>
  </ul>
  <div
    className="c8"
    onClick={[Function]}
  >
    <p
      className="c9"
    >
      Svg file : 
      <br />
      Store_StoreLocation--Colors_48px.svg
    </p>
  </div>
</li>
`;

exports[`<IconTile /> render corectly monochrom icon 1`] = `
.c7 {
  width: 100%;
  padding: 3px 5px;
  font-size: 10px;
  color: #999;
  line-height: 2em;
  cursor: pointer;
  position: relative;
  box-sizing: border-box;
}

.c7:not(:last-child) {
  border-bottom: solid 1px #f5f5f5;
}

.c0 {
  -webkit-flex: 0 1 30.8%;
  -ms-flex: 0 1 30.8%;
  flex: 0 1 30.8%;
  list-style: none;
  border: 1px solid #eee;
  box-sizing: border-box;
  border-radius: 4px;
  margin: 0.5rem;
}

.c2 {
  text-align: center;
  height: 104px;
  line-height: 3.75;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 img,
.c2 svg {
  margin: auto;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 !important;
  border-top: 1px solid #eee;
  border-bottom: 1px solid #eee;
}

.c4 {
  -webkit-flex: 0 1 20%;
  -ms-flex: 0 1 20%;
  flex: 0 1 20%;
  list-style: none;
  text-align: center;
  font-size: 0.65rem;
  line-height: 1.62rem;
  background: #f2f2f2;
  color: darkgrey;
}

.c4:not(:first-child) {
  border-left: 1px solid #eee;
}

.c5 {
  -webkit-flex: 0 1 20%;
  -ms-flex: 0 1 20%;
  flex: 0 1 20%;
  list-style: none;
  text-align: center;
  font-size: 0.65rem;
  line-height: 1.62rem;
  background: #f2f2f2;
  color: darkgrey;
  background: #78be20;
  color: white !important;
}

.c5:not(:first-child) {
  border-left: 1px solid #eee;
}

.c6 {
  box-shadow: none;
  margin: 0;
  padding: 0;
  border: none;
  font-size: 0.65rem;
  line-height: 1;
  background: none;
  cursor: pointer;
  color: inherit;
  padding: 0.5rem 0.25rem;
}

.c1 {
  padding: 1rem;
  border-bottom: 1px solid #eee;
  font-size: 0.75rem;
  margin: 0;
  font-weight: 600;
  word-break: break-all;
  height: 40px;
}

.c8 {
  padding: 0 3px;
  margin: 3px;
  line-height: 1.25;
  display: block;
  word-break: break-all;
  position: relative;
}

<li
  className="c0"
>
  <p
    className="c1"
  >
    Mobile
  </p>
  <div
    className="c2"
  >
    <svg
      fill="#454545"
      height="48px"
      viewBox="0 0 48 48"
      width="48px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M32 7H16a2 2 0 00-2 2v30a2 2 0 002 2h16a2 2 0 002-2V9a2 2 0 00-2-2zm-1 3v22H17V10zM17 38v-4h14v4z"
      />
      <path
        d="M25 35h-2a1 1 0 000 2h2a1 1 0 000-2z"
      />
    </svg>
  </div>
  <ul
    className="c3"
  >
    <li
      className="c4"
    >
      16px
    </li>
    <li
      className="c4"
    >
      24px
    </li>
    <li
      className="c4"
    >
      32px
    </li>
    <li
      className="c5"
    >
      <button
        className="c6"
        onClick={[Function]}
        type="button"
      >
        48px
      </button>
    </li>
    <li
      className="c4"
    >
      64px
    </li>
  </ul>
  <div
    className="c7"
    onClick={[Function]}
  >
    <p
      className="c8"
    >
      React/vue : 
      <br />
      Mobile48
    </p>
  </div>
  <div
    className="c7"
    onClick={[Function]}
  >
    <p
      className="c8"
    >
      Svg file : 
      <br />
      Device_Mobile_48px.svg
    </p>
  </div>
</li>
`;
